packages/builder/src/components/design/settings/controls/ButtonActionEditor/actions/DownloadFile.svelte

```diff
@@ -0,0 +1,110 @@
+<script>
+  import { Select, Label } from "@budibase/bbui"
+  import { onMount } from "svelte"
+  import DrawerBindableInput from "components/common/bindings/DrawerBindableInput.svelte"
+  import { FieldType } from "@budibase/types"
+  import { tables, viewsV2 } from "stores/builder"
+
+  export let parameters
+  export let bindings = []
+
+  const fileOptions = [
+    {
+      label: "Attachment",
+      value: "attachment",
+    },
+    {
+      label: "URL",
+      value: "url",
+    },
+  ]
+
+  $: tableOptions = $tables.list.map(table => ({
+    label: table.name,
+    resourceId: table._id,
+    schema: table.schema,
+  }))
+  $: viewOptions = $viewsV2.list.map(view => ({
+    label: view.name,
+    resourceId: view.id,
+    schema: view.schema,
+  }))
+  $: options = [...(tableOptions || []), ...(viewOptions || [])]
+
+  $: selectedTable =
+    parameters.tableId && options.find(t => t.resourceId === parameters.tableId)
+  $: attachmentColumns =
+    selectedTable &&
+    Object.values(selectedTable.schema).filter(c =>
+      [FieldType.ATTACHMENTS, FieldType.ATTACHMENT_SINGLE].includes(c.type)
+    )
+
+  onMount(() => {
+    if (!parameters.type) {
+      parameters.type = "attachment"
+    }
+  })
+</script>
+
+<div class="root">
+  <Label small>File</Label>
+  <Select
+    placeholder={null}
+    bind:value={parameters.type}
+    options={fileOptions}
+  />
+  {#if parameters.type === "attachment"}
+    <Label>Table</Label>
+    <Select
+      placeholder={null}
+      bind:value={parameters.tableId}
+      {options}
+      getOptionLabel={table => table.label}
+      getOptionValue={table => table.resourceId}
+    />
```

Comment:
Rather than have 3 components here that all need configured, we could simply filter the binding list down to only inlude attachment columns. This is what I was originally thinking.

If you do
```js
bindings.filter(x => x.fieldSchema?.type === "attachment")
```
Then you can just have a single picker, with the values being the `runtimeBinding`s of those bindings. Then it's much easier for the user to configure 👍

Reply:
We can't do this sadly - due to the way zipping has to work. If the user passes in a bunch of different attachments we don't know what table/row the attachments came from. We played around with different ways to try and convert the attachment list back to the information we actually need, but it was super problematic, often requiring extra metadata on the rows.

The only other way would be to download all the files somehow into memory in the browser and then zip it there, but this proved way more complex than allowing the backend to build a zip out of all the attachments.

The other option was to just download all of the files individually, which we felt was a worse experience than it being zipped when there are multiple.

Reply:
We tried the egress enrichment - it can sort of work but the use of it is very odd - as we want to download a single row's attachment column, so we have to do something like `attachments[0].tableId` etc - it assumes that if a binding array is passed in we just download whatever is in the first location.

Issue with this is an edge case, but one that cannot be fixed, if a user manages to get attachments from two columns together here then they'll only get one. This in combination with how odd the code looked and how awful the enrichment was going to be, it felt better to just get the information we need from the user directly.

Reply:
Cool, no problem.

Reply:
Just to finalise this as well, we did consider another method of building an endpoint that takes a list of attachment URLs and just downloads them, however its very difficult to secure an endpoint like this, it would be very difficult for us to know if the user actually has access to the row which the attachment came from, without having to search by the attachment URL and then working out if the user can access all of the attachments - again rather than going down this route we preferred just getting the info from the user.